virt: differentiate Amazon EC2 bare-metal from VM instances using SMBIOS data

nmeyerhans opened the pull request:
Please note, this builds on #17315 and includes all of its commits. cc @bjacquin

#17315 introduces the ability to detect and recognize Amazon EC2 Nitro-based instances. However, the kernel does not expose enough formatted DMI information to allow us to distinguish between VM and bare metal instances. This is not an issue on x86, where KVM overrides the CPUID CPU instruction with a magic string "KVMKVMKVM", the presence of which indicates a VM and the lack of which indicates bare-metal. However, on arm64, there is no equivalent hint from KVM, and we need to look elsewhere. (See #18929)

The [SMBIOS specification](https://www.dmtf.org/sites/default/files/standards/documents/DSP0134_3.4.0.pdf) provides a mechanism in the BIOS Information Table (section 7.1.2.2) whereby a firmware implementation can indicate that the SMBIOS table describes a virtual machine. On EC2 instances, this bit is set on VMs, and empty on bare-metal instances. The contents of the BIOS Information table are exposed via the kernel via the `/sys/firmware/dmi/entries` interface, so we look there and set the virtualization state accordingly.

`systemd-detect-virt` shows the following:

**EC2 arm64 VM**
```
# Before this change
admin@ip-10-0-0-12:~$ SYSTEMD_LOG_LEVEL=debug systemd-detect-virt
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
No virtualization found in DMI
UML virtualization not found in /proc/cpuinfo.
No virtualization found in CPUID
Virtualization XEN not found, /proc/xen does not exist
/proc/device-tree: No such file or directory
This platform does not support /proc/sysinfo
Found VM virtualization none
none

# With changes from #17315
admin@ip-10-0-0-178:~$ sudo sh -c "SYSTEMD_LOG_LEVEL=debug LD_LIBRARY_PATH=$PWD ./systemd-detect-virt"
Found container virtualization none.
Virtualization Amazon EC2 found in DMI (/sys/class/dmi/id/sys_vendor)
Found VM virtualization amazon
amazon

# With this change
admin@ip-10-0-0-12:~/systemd$ sudo sh -c "SYSTEMD_LOG_LEVEL=debug LD_LIBRARY_PATH=$PWD/foo/src/shared/ ./foo/systemd-detect-virt"
Found container virtualization none.
Virtualization Amazon EC2 found in DMI (/sys/class/dmi/id/sys_vendor)
DMI BIOS Extension table indicates virtualization
Found VM virtualization amazon
amazon
```

**EC2 arm64 bare-metal**
```
# The current state
admin@ip-10-0-0-170:~/systemd$ SYSTEMD_LOG_LEVEL=debug systemd-detect-virt
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
No virtualization found in DMI
UML virtualization not found in /proc/cpuinfo.
No virtualization found in CPUID
Virtualization XEN not found, /proc/xen does not exist
/proc/device-tree: No such file or directory
This platform does not support /proc/sysinfo
Found VM virtualization none
none

# With only the changes from #17315 (incorrect)
admin@ip-10-0-0-211:~/systemd$ sudo sh -c "SYSTEMD_LOG_LEVEL=debug LD_LIBRARY_PATH=$PWD/foo/src/shared/ ./foo/systemd-detect-virt"
Found container virtualization none.
Virtualization Amazon EC2 found in DMI (/sys/class/dmi/id/sys_vendor)
Found VM virtualization amazon
amazon

# With this change
admin@ip-10-0-0-211:~/systemd$ sudo sh -c "SYSTEMD_LOG_LEVEL=debug LD_LIBRARY_PATH=$PWD/foo/src/shared/ ./foo/systemd-detect-virt"
Found container virtualization none.
Virtualization Amazon EC2 found in DMI (/sys/class/dmi/id/sys_vendor)
DMI BIOS Extension table does not indicate virtualization
No virtualization found in CPUID
Virtualization XEN not found, /proc/xen does not exist
/proc/device-tree: No such file or directory
UML virtualization not found in /proc/cpuinfo.
This platform does not support /proc/sysinfo
Found VM virtualization none
none
```

It's worth reiterating the issues that @ozbenh [pointed out](https://github.com/systemd/systemd/issues/18929#issuecomment-823720725) in #18929. Specifically that this approach requires root access and presence of the dmi-sysfs.ko kernel module. If either of these is not available, then arm64 bare-metal instances will be incorrectly identified as virtual machines. However, given current kernel interfaces, there is no alternative.

bluca commented:
Needs a rebase, also there are several styling nits to solve.

Most importantly, as it was mentioned in the original PR, if the hypervisor is KVM shouldn't it just be detected as KVM? This detection system doesn't differentiate between vendors - eg, you get HyperV rather than Azure/Microsoft and so on.

bluca commented on src/basic/virt.c:
```diff
@@ -180,8 +179,43 @@ static int detect_vm_dmi(void) {
+
+        int r;
+        r = detect_vm_dmi_vendor();
+        if(r == VIRTUALIZATION_AMAZON) {
```
space between `if` and `(`

nmeyerhans replied:
done

bluca commented on src/basic/virt.c:
```diff
@@ -180,8 +179,43 @@ static int detect_vm_dmi(void) {
+        r = detect_vm_dmi_vendor();
+        if(r == VIRTUALIZATION_AMAZON) {
+                /* The DMI vendor tables in /sys/class/dmi/id don't help us distinguish between Amazon EC2
+                   virtual machines and bare-metal instances, so we need to look deeper.  The EC2 firmware
```
Multi line comments in our code base are usually in the form:

```
/*
 *
 * .. */
 ```

nmeyerhans replied:
fixed

bluca commented on src/basic/virt.c:
```diff
@@ -180,8 +179,43 @@ static int detect_vm_dmi(void) {
+                   exposed via the kernel in /sys/firmware/dmi/entries/0-0. */
+                _cleanup_free_ char *s = NULL;
+                size_t readsize;
+                // int i = read_one_line_file("/sys/firmware/dmi/entries/0-0/raw", &s);
```
leftover?

nmeyerhans replied:
yes, deleted.

bluca commented on src/core/machine-id-setup.c:
```diff
@@ -58,21 +58,23 @@ static int generate_machine_id(const char *root, sd_id128_t *ret) {
-                        if (id128_read("/sys/firmware/devicetree/base/vm,uuid", ID128_UUID, ret) >= 0) {
-                                log_info("Initializing machine ID from KVM UUID.");
-                                return 0;
+                } else {
```
```
} else if (IN_SET(detect_vm(), VIRTUALIZATION_KVM, VIRTUALIZATION_AMAZON)) {
```

nmeyerhans replied:
fixed

bluca commented on src/basic/virt.c:
```diff
@@ -180,8 +179,43 @@ static int detect_vm_dmi(void) {
+                _cleanup_free_ char *s = NULL;
+                size_t readsize;
+                // int i = read_one_line_file("/sys/firmware/dmi/entries/0-0/raw", &s);
+                int i = read_full_virtual_file("/sys/firmware/dmi/entries/0-0/raw", &s, &readsize);
```
we usually declare variable at the top of the innermost block where they are used

nmeyerhans replied:
fixed

nmeyerhans commented:
> Needs a rebase, also there are several styling nits to solve.

Thanks for the review! I will push an update to address the nits shortly.

> Most importantly, as it was mentioned in the original PR, if the hypervisor is KVM shouldn't it just be detected as KVM? This detection system doesn't differentiate between vendors - eg, you get HyperV rather than Azure/Microsoft and so on.

I don't think so. I went in to the reasons for this in some detail in [a comment](https://github.com/systemd/systemd/issues/18929#issuecomment-797007023) on #18929. The short answer is, there's no way to detect the presence of KVM non non-x86 platforms. Further, I'd argue that detecting KVM by itself isn't particularly meaningful by itself. Other than the lowest-level resources, all the interesting stuff you see in a KVM system (firmware, devices, etc) comes from outside of KVM, e.g. QEMU, crosvm, Firecracker, or others. This includes all the strings you find in /sys/class/dmi/id/ and other DMI data. Even in the case where /sys/class/dmi/id/product_name contains "KVM Virtual Machine", that value is populated by QEMU and only when running under KVM acceleration. So I'd argue that systemd should still probably report QEMU as the hypervisor in that case. (Maybe as "qemu+kvm" or something like that to recognize that KVM is providing acceleration to QEMU.)

nmeyerhans commented:
Rebased against main as of 4990780ed01bff58e6fa147f8e6f2cb7a568bf8f

ozbenh commented:
> > Most importantly, as it was mentioned in the original PR, if the hypervisor is KVM shouldn't it just be detected as KVM? This detection system doesn't differentiate between vendors - eg, you get HyperV rather than Azure/Microsoft and so on.
> 
> I don't think so. I went in to the reasons for this in some detail in [a comment](https://github.com/systemd/systemd/issues/18929#issuecomment-797007023) on #18929. 

More specifically, I think it's worth quoting the following from Lennart: "I figure it would make sense to declare (and document) that what we report as "kvm" actually means "kvm+qemu", and then treat the aws thing as a totally different thing".

In our case, the virtual environment is provided by Amazon custom userspace, which does use KVM for the low level CPU virtualization but provides an otherwise different runtime environment. It does make sense in this case to treat it as a separate hypervisor. So would be true of firecracker btw, it should probably be its own thing as well.

Noah: Do you want to be more precise at calling this VIRTUALIZATION_AMAZON_EC2 (or NITRO or something like this...) to differenciate from Xen ?

bluca commented on src/basic/virt.c:
```diff
@@ -1,4 +1,4 @@
-/* SPDX-License-Identifier: LGPL-2.1-or-later */
+/* SPDX-License-Identifier: LGPL-2.1ror-later */
```
Spurious change?

nmeyerhans replied:
fixed.

bluca commented:
```
In file included from ../src/basic/macro.h:491,
                 from ../src/basic/alloc-util.h:9,
                 from ../src/basic/virt.c:11:
../src/basic/virt.c: In function â€˜detect_vm_dmiâ€™:
../src/basic/virt.c:206:35: error: format â€˜%luâ€™ expects argument of type â€˜long unsigned intâ€™, but argument 7 has type â€˜size_tâ€™ {aka â€˜unsigned intâ€™} [-Werror=format=]
  206 |                         log_debug("Only read %lu bytes from /sys/firmware/dmi/entries/0-0/raw", readsize);
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~  ~~~~~~~~
      |                                                                                                 |
      |                                                                                                 size_t {aka unsigned int}
../src/basic/log.h:196:86: note: in definition of macro â€˜log_full_errno_zerookâ€™
  196 |                         ? log_internal(_level, _e, PROJECT_FILE, __LINE__, __func__, __VA_ARGS__) \
      |                                                                                      ^~~~~~~~~~~
../src/basic/log.h:224:28: note: in expansion of macro â€˜log_fullâ€™
  224 | #define log_debug(...)     log_full(LOG_DEBUG,   __VA_ARGS__)
      |                            ^~~~~~~~
../src/basic/virt.c:206:25: note: in expansion of macro â€˜log_debugâ€™
  206 |                         log_debug("Only read %lu bytes from /sys/firmware/dmi/entries/0-0/raw", readsize);
      |                         ^~~~~~~~~
../src/basic/virt.c:206:48: note: format string is defined here
  206 |                         log_debug("Only read %lu bytes from /sys/firmware/dmi/entries/0-0/raw", readsize);
      |                                              ~~^
      |                                                |
      |                                                long unsigned int
```

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+
+        int r;
+        r = detect_vm_dmi_vendor();
+        if (r == VIRTUALIZATION_AMAZON) {
```
hmm, why check vendor *and* this bit? To distuingish amazon physical machines and amazon VMs?

nmeyerhans replied:
Yes, we need both checks. The DMI vendor information is identical between VMs and physical systems.

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+                _cleanup_free_ char *s = NULL;
+                int i;
+                size_t readsize;
+                i = read_full_virtual_file("/sys/firmware/dmi/entries/0-0/raw", &s, &readsize);
```
we almost always use `r` as the variable to store function result/error codes in. Please do so here too, to follow our usual style. i.e. drop "int i", and just use the "r" you already have

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+                size_t readsize;
+                i = read_full_virtual_file("/sys/firmware/dmi/entries/0-0/raw", &s, &readsize);
+                if (i < 0) {
+                        log_debug("Unable to read /sys/firmware/dmi/entries/0-0/raw: %d", i);
```
return log_debug_errno(r, "…: %m");

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+                        return r;
+                }
+                if (readsize < 20) {
+                        log_debug("Only read %lu bytes from /sys/firmware/dmi/entries/0-0/raw", readsize);
```
%zu is the right format string for size_t values. Please say `(expected 20)` or so in the message.

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+                        log_debug("Only read %lu bytes from /sys/firmware/dmi/entries/0-0/raw", readsize);
+                        return r;
+                }
+                unsigned byte = s[19];
```
type confusion. s is of type char*, you assign it to an `unsigned`. This is an implicit sign cast to a bigger type. Please make this explicit:

```c
uint8_t byte = (uint8_t) s[19];
```

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+                        return r;
+                }
+                unsigned byte = s[19];
+                if (byte ^ 1<<4) {
```
i must admit i don't know C operator precedence well enought to be able to say whether `^` or `<<` binds more strongly. please use `()` to make this clear to the casual onlooker.

nmeyerhans replied:
The logic is changed in the latest version, but the structure is the same. I've added the parens for clarity.

poettering commented on src/basic/virt.c:
```diff
@@ -181,8 +179,43 @@ static int detect_vm_dmi(void) {
+                        return VIRTUALIZATION_NONE;
+                }
+                log_debug("DMI BIOS Extension table indicates virtualization");
+        }
```
I'd prefer if the check for this bit could be split out into a function of its own, it's entirely generic and not specific to amazon after all. Morevoer, I think we should always call this at the end of our tests as a last-ditch effort to detect VM environments we otherwise don't recognize. And then return `VIRTUALIZATION_VM_OTHER`, to indicate that yes, we know we are in a VM, but no we have no clue in what.

nmeyerhans replied:
Are you sure you want to trust everybody's firmware to set that bit correctly?

nmeyerhans replied:
The latest revision moves this into a separate function. It's called in two places:
1. After the `/sys/class/dmi/id` checks identify an Amazon EC2 system, to validate that it is a VM rather than a bare-metal instance.
2. If no virtualization is detected, with return `VIRTUALIZATION_VM_OTHER` if this bit is set. I still have some concern that this might lead to hardware systems with buggy firmware reporting as VMs, but I don't know of any specific instances of such hardware.

poettering replied:
> Are you sure you want to trust everybody's firmware to set that bit correctly?

Well, given that the flag is kinda hidden in some auxiliary struct that mainboard vendors hopefully never bother with I'd at least try it. If it doesnt work, and we get reports of false positives, we can drop it again. But the only way to figure it out is to try.

nmeyerhans commented on man/systemd-detect-virt.xml:
```diff
@@ -72,6 +72,11 @@
             <entry>Linux KVM kernel virtual machine, with whatever software, except Oracle Virtualbox</entry>
```
updated.

nmeyerhans commented on src/shared/machine-id-setup.c:
```diff
@@ -59,9 +59,7 @@ static int generate_machine_id(const char *root, sd_id128_t *ret) {
-
-                } else if (detect_vm() == VIRTUALIZATION_KVM) {
-
+                } else if (IN_SET(detect_vm(), VIRTUALIZATION_KVM, VIRTUALIZATION_AMAZON)) {
```
> does amazon fill in the DMI product ID with the machine UUID?

Yes.

poettering replied:
please update the comment below to say this clearly

poettering replied:
hmm, this was never adressed?

nmeyerhans replied:
Actually, was that check right before? It was only making use of the `-uuid` parameter for `VIRTUALIZATION_KVM` systems, but that parameter is supported by QEMU regardless of whether KVM is used. I'm going to propose updating it (and the comments) to:
`IN_SET(detect_vm(), VIRTUALIZATION_KVM, VIRTUALIZATION_AMAZON, VIRTUALIZATION_QEMU)`

poettering commented on man/systemd-detect-virt.xml:
```diff
@@ -69,7 +69,12 @@
+
+          <row>
+            <entry><varname>amazon</varname></entry>
+            <entry>Amazon EC2 Nitro</entry>
```
maybe add "(using the KVM interfaces)", because we say so for the virtualbox case

nmeyerhans replied:
Changed the text to "Amazon EC2 Nitro using Linux KVM"

poettering commented on src/basic/virt.c:
```diff
@@ -180,8 +187,60 @@ static int detect_vm_dmi(void) {
+        int r;
+        r = read_full_virtual_file("/sys/firmware/dmi/entries/0-0/raw", &s, &readsize);
+        if (r < 0) {
+                log_debug_errno(r, "Unable to read /sys/firmware/dmi/entries/0-0/raw: %m");
```
we usually end messages like this with with `, ignoring:`, to clarify this is a message we knowingly ignore.

i.e. change to:

```c
log_debug_errno(r, "Unable to read /sys/firmware/dmi/entries/0-0/raw, ignoring: %m");
```

nmeyerhans replied:
done

poettering commented on src/basic/virt.c:
```diff
@@ -180,8 +187,60 @@ static int detect_vm_dmi(void) {
+                return SMBIOS_VM_BIT_UNKNOWN;
+        }
+        uint8_t byte = (uint8_t) s[19];
+        if (byte & (1<<4)) {
```
should probably use 1U

nmeyerhans replied:
done

poettering commented:
looks excellent. Only nitpicks left.

(anothr nitpick: i'd put single empy lines between some of your very densely packed codeblocks, but it doesnt matter)

nmeyerhans commented:
> 
> (anothr nitpick: i'd put single empy lines between some of your very densely packed codeblocks, but it doesnt matter)

No problem. I've added some vertical whitespace to delineate some of the logical blocks.

poettering commented:
the ubuntu s390 build fails:

```
../src/basic/virt.c: In function ‘detect_vm_smbios’:
../src/basic/virt.c:227:16: error: ‘SMBIOS_VM_BIT_UNKNOWN’ undeclared (first use in this function)
         return SMBIOS_VM_BIT_UNKNOWN;
                ^~~~~~~~~~~~~~~~~~~~~
../src/basic/virt.c:227:16: note: each undeclared identifier is reported only once for each function it appears in
```

probably something wrong with the ifdeffery

poettering commented on src/basic/virt.c:
```diff
@@ -21,6 +21,14 @@
 #include "string-util.h"
 #include "virt.h"
 
+#if defined(__i386__) || defined(__x86_64__) || defined(__arm__) || defined(__aarch64__)
```
this conditionalization must be dropped, you use the enum further down in all cases

nmeyerhans replied:
Whoops. Sorry about that. Fixed.

poettering commented:
Hmm, I figure the whole detect_vm_dmi_vendor() and detect_vm_dmi_smbios() need to be ifdeffed out on those non-DMI archs. right now you ifdef out the body of one and leave the other in there entirely.

poettering commented:
also needs a rebase now. sorry!

nmeyerhans commented on src/basic/virt.c:
```diff
@@ -181,8 +185,64 @@ static int detect_vm_dmi(void) {
+        if (r == VIRTUALIZATION_NONE && detect_vm_smbios() == SMBIOS_VM_BIT_SET)
+                return VIRTUALIZATION_VM_OTHER;
+        return r;
+#endif
```
Yes, you're right; as is, with the conditional code enabled, there will be a small unreachable block below the return on line 244.

nmeyerhans commented:
> Hmm, I figure the whole detect_vm_dmi_vendor() and detect_vm_dmi_smbios() need to be ifdeffed out on those non-DMI archs. right now you ifdef out the body of one and leave the other in there entirely.

Yep. In the latest revision I'll move the entire definition of both `detect_vm_dmi_vendor()` and `detect_vm_smbios()` into an ifdef block, since the only calls to those functions occur in a similar block.

> also needs a rebase now. sorry!

done.

poettering commented:
hmm, the commits are strangely split up. i.e. we genreally want PRs that are "clean" in the sense that they don't fix up bugs that were introduced in an earlier commit of the same PR. The spdx lincense tag thing however was introduced in the first patch and fixed in a later one.

We are keen on this to keep things nicely bisectable: every commit should reflect a logical step (even though this means not a historical one, necessarily), and stand on its own feet.

There's also something wrong with the docs changes: the first commit carries some of them, and the last then change them again?

it would also be ok to squash all three patches into one btw, if that simplifies things.

also, https://github.com/systemd/systemd/pull/19403#discussion_r621590553 wasnt addressed so far

nmeyerhans commented:
> hmm, the commits are strangely split up. i.e. we genreally want PRs that are "clean" in the sense that they don't fix up bugs that were introduced in an earlier commit of the same PR. The spdx lincense tag thing however was introduced in the first patch and fixed in a later one.

Yep, sorry. I addressed some of your comments in the top commit when they should have been in earlier commits. Should be fixed now.

> also, [#19403 (comment)](https://github.com/systemd/systemd/pull/19403#discussion_r621590553) wasnt addressed so far

I've addressed that, as well. I also added `VIRTUALIZATION_QEMU` to the set, since the `-uuid` flag can be provided to qemu independent of whether or now KVM is in use. Let me know if you disagree with that change for some reason.

poettering commented:
excellent! thanks! lgtm
